### ISSUE TITLE
Clarification on the relationship between num_unroll_steps and infer_context_length in UniZero

### ISSUE THREAD

**Issue by Tiikara** [opendilab/LightZero#248]
I've been studying the UniZero implementation and I have a question about two key parameters:

```python
num_unroll_steps = 10
infer_context_length = 4
```

I noticed that these values are different, and I'm curious about the reasoning behind this design choice. Specifically:

1. Why is `num_unroll_steps` set higher than `infer_context_length`?
2. Does this configuration affect the training or inference efficiency?
3. Would there be any drawbacks to setting these values equal to each other?

I'd greatly appreciate any insights you could provide on the rationale behind these parameter choices and their impact on the model's performance and behavior.

Thank you for your time and for creating this interesting algorithm.

**Comment by puyuan1996** [comment]
- Hello, thank you very much for your attention.
- `num_unroll_steps` refers to the `context_length` during training (i.e., the length of the training sequence). This naming is to keep consistent with the MuZero algorithm. Indeed, we can consider changing it to `train_context_length` in the future. `infer_context_length` represents the `context_length` during collect/eval (i.e., only the most recent length of `KVCache` is retained for inference).
- The reason we set `num_unroll_steps` larger than `infer_context_length` is that in our current implementation, the positional encoding uses the classic [nn.Embedding](https://github.com/opendilab/LightZero/blob/main/lzero/model/unizero_world_models/world_model.py#L59), rather than the latest advanced positional encodings, such as RoPE, so the extrapolation ability is limited. Meanwhile, we hope the model sees a longer context during training (referencing this [paper](https://openreview.net/forum?id=agPpmEgf8C), predicting further into the future may help form some global structure) and select `infer_context_length` based on the task characteristics during inference. In our implementation, we set `infer_context_length = 4` because generally, stacking 4 frames in Atari tasks is sufficient for optimal decision-making. Both longer `num_unroll_steps` and `infer_context_length` will require more CUDA memory and potentially perform better, so we need to decide their values based on the task characteristics and compute cost, and we can consider some adaptive method in the future.
- Thank you again for your attention, and feel free to raise any questions or suggestions at any time.

**Comment by Tiikara** [comment]
Thank you very much for your response. I would like to know more - I tried running unizero on my own with default parameters, and the envstep graphs from tensorboard are not the same as in the paper. From the paper, it's clear that when envstep equals 0.1M, the reward starts to increase. I would like to know what envstep means in the paper?

<img width="616" alt="image" src="https://github.com/user-attachments/assets/515a4f2f-dd16-4eca-bb87-8403f1238a95">

**Comment by puyuan1996** [comment]
The `EnvSteps` in the paper refer to the total number of interaction steps with the environment. The curve data in our paper comes from pre-reconstruction experiments, while the code in the main branch has been refactored and optimized. It is reasonable for there to be slight differences in the curves, but there should not be significant discrepancies. Please ensure you are using the default configuration ([atari_unizero_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_config.py) or [atari_unizero_stack4_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_stack4_config.py)). Perhaps you could try running [atari_unizero_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_config.py) once to see the effect and provide complete training tensorboard logs for us to analyze the cause. Thank you.

**Comment by Tiikara** [comment]
I have conducted a thorough comparison between the parameters specified in the paper and those in the configuration files. Unfortunately, I did not identify any significant discrepancies. I executed atari_unizero_stack4_config.py with only minor modifications: I adjusted the frequency of checkpoint saves and enabled video file recording. These changes should not materially affect the training process or results.

[formatted_total_config.zip](https://github.com/user-attachments/files/16367157/formatted_total_config.zip)

**Comment by puyuan1996** [comment]
Thank you for your feedback. We will verify and address the performance issues related to [atari_unizero_stack4_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_stack4_config.py) within a week. The code using stack1 in the main branch, [atari_unizero_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_config.py), has already been confirmed to perform consistently with the curves presented in the paper. We recommend using [atari_unizero_config.py](https://github.com/opendilab/LightZero/blob/main/zoo/atari/config/atari_unizero_config.py) for your tests and research in the meantime. Thank you for your patience.

**Comment by Tiikara** [comment]
Thank you for your prompt and helpful reply. I truly appreciate your dedication to addressing this issue. The work you're doing with LightZero is impressive and valuable to the research community.

**Comment by Tiikara** [comment]
or clarification: In the Tensorboard, do the collector_step and tabs with the *_step prefix represent envstep? Or do they represent *_iter? I want to ensure I'm interpreting the data correctly. Thank you for your assistance.

**Comment by puyuan1996** [comment]
Hello, in Tensorboard, tags with the `_step` prefix indicate the `environment steps` (EnvStep), while tags with the `_iter` prefix indicate the `training iterations` (train iteration). For detailed information, you can refer to this [documentation](https://opendilab.github.io/LightZero/tutorials/logs/logs.html). Unless otherwise stated, the learning curves in the paper generally refer to the curves in the evaluator. Best wishes.

**Comment by Tiikara** [comment]
Thank you very much for your clear and informative response. I believe this addresses all my current questions, so I'll be closing this thread. Thank you again for your time and support throughout this discussion.
